### PR TITLE
spec: allow named endpoints

### PR DESCRIPTION
**Pull request by LinusU**
See discussion in #196

### CODE REVIEW

**Comment by zdne** [comment]
When I think about it more conceptually – the endpoint as such is actually an action section in API Blueprint parlance (albeit it implies a new resource). Therefore it should be defined under Action and not resource as such. 

This, however, has some implications – suddenly we can have "action" section at the top of a blueprint so the "Blueprint document structure" needs to be updated accordingly. 

Internally, in the parser, this should be still picked up by Resource parser as it essentially creates a new resource (in AST) but this is opaque to user. 

Furthermore parser should not warn when an action with the same URI template (but different method) is described, instead, it should add it under the respective resource. 

Overall as simply as it looks it is actually quite complex problem :)

**Comment by LinusU** [comment]
Hmm, I guess that in that case we'd want to change the behaviour of `# GET /e1` as well, no?

I'm not sure if I'm happy with letting content "jump" to different places thought, it would make the internals quite a bit more complicated?

It would also make it complicated to do "guides" with blueprint since the content might not necessarily come in the order that you specify it in the markdown.

e.g.
```markdown
# Create a user [POST /users]

...

# Log in as the user [POST /sessions]

...

# List all users [GET /users]

...
```

This would now list `List all users` before it listed `Log in as the user` making that call fail...

File: examples/13. Named Endpoints.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+# Group Quick start
+
+## Create user [POST /users]
+
+The first thing we need to do is to add a user that we can later log in as. Do
+this by sending a POST request with the relevant details.
+
++ Request
+  + Headers
+    Content-Type: application/json
```
Comment (pksunkara) [inline examples/13. Named Endpoints.md]: The content inside `Headers` and `Body` list items should be code blocks. So, indent them by 12 spaces.

File: examples/13. Named Endpoints.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+# Group Quick start
+
+## Create user [POST /users]
+
+The first thing we need to do is to add a user that we can later log in as. Do
+this by sending a POST request with the relevant details.
+
++ Request
+  + Headers
```
Comment (pksunkara) [inline examples/13. Named Endpoints.md]: The second-level list items should be indented by 4 spaces according to our API Blueprint examples style guide.

**Comment by pksunkara** [comment]
A few styling issues other than looks good. But I agree with @zdne that even though this issue looks simple, it is a bit complex one. We need to do some thinking on this.

File: API Blueprint Specification.md [hunk 3]
```diff
@@ -534,7 +534,13 @@ Defined by an [HTTP request method][httpmethods] followed by [URI template][urit
 
     # <HTTP request method> <URI template>
 
-> **NOTE:** In this case the rest of this section represents the [Action section](#def-action-section) including its description and nested sections and **follows the rules of Action section instead**.
+**-- or --**
+
+Defined by a resource [name (identifier)](#def-identifier) followed by an [HTTP request method][httpmethods] and an [URI template][uritemplate] enclosed in square brackets `[]`:
+
+    # <identifier> [<HTTP request method> <URI template>]
```
Comment (pksunkara) [inline API Blueprint Specification.md]: By the way, if we are doing this, please add some info about this to the section title `Header Keywords`.

**Comment by zdne** [comment]
Hey @LinusU I was thinking about it a bit more and I think you are indeed right with: 

> I guess that in that case we'd want to change the behaviour of # GET /e1 as well, no?

Now doing that and also addressing 

>  the content might not necessarily come in the order that you specify it in the markdown.

is quite radical change... So lets recap on what are the options:

## 1. Add name to an endpoint

Simply allow adding a name to `# <HTTP request method> <URI template>` action and leave everything else as is. 

### Pros
- simplest way to add a name
- doesn't change current behavior

### Cons
- everything stays under one resource - you will loose the order of these "named endpoints" in the case when they are hitting the same URI

## 2. Separate endpoints from resources

Introduce a new top-level "endpoint" section – for `# <HTTP request method> <URI template>` or `# <Name> [<HTTP request method> <URI template>]`. 

### Pros
- maintain order of endpoints
- clearly distinguish between resource and endpoint architecture (or document structure)

### Cons
- need to change AST
- changing behavior of existing things
- much more complex on implementation side
- affects all the tools

---

Thoughts ?

**Comment by LinusU** [comment]
@zdne great writeup.

Number 1 is what I have implemented in this pull request, the only change really is to allow a name to the resource that gets generated.

I do like number 2 quite much thought, but as you say it's much more work involved. I think that we need to clarify a bit an endpoint is. Is it a path (e.g. `/cars`) or a method + path (e.g. `GET /cars`)? Would an endpoint contain an action? Or maybe any number of action? Or is the endpoint in itself an action?

If we try and work our way from the ast, which I think is a good approach, how would the new ast structure look?

I guess that it looks something like this currently:

- `category` *root*
  - `resource`
    - `action`
  - `category`
    - `resource`
      - `action`

And it would become something like this if we added endpoints:

- `category` *root*
  - `endpoint`
  - `resource`
    - `action`
  - `category`
    - `endpoint`
    - `resource`
      - `action`

Or maybe `resource` really just should be a `category` and `action` should be an `endpoint`. The uri would than be moved and placed in every `endpoint` instead of the one `resource`.

- `category` *root*
  - `endpoint`
  - `category`
    - `endpoint`

Very simple ast at least. Every endpoint should then contain `uriTemplate` and `method`.

Maybe a good approach would be to implement number 1 now since that dosen't change anything for the existing ecosystem, and then try and figure out what we think about number 2.

I really like the ideas in number 2 thought, it makes it possible to specify APIs that aren't RESTful which I think is important.

**Comment by zdne** [comment]
@LinusU OK so let's break it into two parts

1. Just adding names to top-level method + URI pairs
2. Changing the internals to reflect ordering

**Comment by zdne** [comment]
> I think that we need to clarify a bit an endpoint is.

:+1: 

In the current context I am using the term "endpoint" for a method + URI pair

>  it makes it possible to specify APIs that aren't RESTful which I think is important.

:+1: 

---

Note @smizell is currently working [Refract Resource Namespace](https://github.com/refractproject/refract-spec/blob/master/namespaces/resource-namespace.md) – this eventually should be the underlaying data structure – I believe he will address these main concept there (resources and transition) using some sort of recursive pattern. 

TL;DR let's figure this one (point 2 as mentioned above) in the Resource namespace first and focus on just adding names to currently un-named method URI pairs.

Sounds good?

File: API Blueprint Specification.md (hunk 3 above)
Comment (zdne) [inline API Blueprint Specification.md]: @pksunkara I think it is OK to leave it as it is, because it is being already covered by "Combinations of an HTTP method and URI Template (e.g. GET /resource/{id})"

File: examples/13. Named Endpoints.md (hunk 2 above)
Comment (zdne) [inline examples/13. Named Endpoints.md]: :+1:

File: examples/13. Named Endpoints.md (hunk 1 above)
Comment (zdne) [inline examples/13. Named Endpoints.md]: :+1:

File: examples/13. Named Endpoints.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    Content-Type: application/json
+  + Body
+    { "username": "Linus", "password": "test" }
```
Comment (zdne) [inline examples/13. Named Endpoints.md]: Consistence: Can we please use the same / similar example data as in the previous examples?

File: examples/13. Named Endpoints.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
++ Response 201
+  + Headers
+    Location: /user/1337
+
+## Log in as the user [POST /session]
+
+We can now log in as the user we created. This is accomplished by creating a new
+session with the relevant login details.
+
++ Request
```
Comment (zdne) [inline examples/13. Named Endpoints.md]: Did you leave Content-Type – `application/json` out on purpose ?

File: examples/13. Named Endpoints.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  + Headers
+    Content-Type: application/json
```
Comment (zdne) [inline examples/13. Named Endpoints.md]: Oh never mind. But I would rather see

```
+ request (application/json)
```

File: examples/13. Named Endpoints.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  + Body
+    { "username": "Linus", "password": "test" }
+
++ Response 201
+  + Headers
```
Comment (zdne) [inline examples/13. Named Endpoints.md]: Indentation

File: examples/13. Named Endpoints.md (hunk 6 above)
Comment (zdne) [inline examples/13. Named Endpoints.md]: doing so you won't need `+ body `

**Comment by zdne** [comment]
@LinusU Please address the review comments on the example and it is good to go! Thanks! :balloon:

**Comment by LinusU** [comment]
@zdne Should be good to go :+1:

File: API Blueprint Specification.md [hunk 8]
```diff
@@ -534,7 +534,13 @@ Defined by an [HTTP request method][httpmethods] followed by [URI template][urit
...
+
+> **NOTE:** In the two latter cases the rest of this section represents the [Action section](#def-action-section) including its description and nested sections and **follows the rules of Action section instead**.
```
Comment (pksunkara) [inline API Blueprint Specification.md]:
```diff
- In the two latter
+ In the latter two
```

**Comment by LinusU** [comment]
@pksunkara Done!

**Comment by pksunkara** [comment]
Looks good. I am toggling this as "Passed Review". :+1:

**Comment by pksunkara** [comment]
@zdne Since I merge apiaryio/snowcrash#332, Do we merge this? Or should we wait for releasing it on Apiary?

**Comment by zdne** [comment]
@pksunkara I think we can merge this / update README (version supported in apiary) and release new version of API Blueprint...

**Comment by pksunkara** [comment]
@LinusU Can you please make a new commit to this PR updating the current format to 1A9 from 1A8? Thanks.

**Comment by LinusU** [comment]
I'm on it!

**Comment by LinusU** [comment]
Done, I did not tag it since it seems that you have always tagged the merge commit :)

**Comment by LinusU** [comment]
Cheers :beers:
